### PR TITLE
Allow only one instance of UX

### PR DESCRIPTION
**Pull request by X9VoiD**

### CODE REVIEW

File: OpenTabletDriver.Plugin/IUXDriver.cs [hunk 1]
```diff
@@ -0,0 +1,9 @@
+using System.Threading.Tasks;
+
+namespace OpenTabletDriver.Plugin
+{
+    public interface IUXDriver
```
Comment (InfinityGhost) [inline OpenTabletDriver.Plugin/IUXDriver.cs:5]: `IUserInterface` makes more sense here.

File: OpenTabletDriver.UX/RPC/UXDriver.cs [hunk 2]
```diff
@@ -0,0 +1,23 @@
+using System.Threading.Tasks;
+using Eto.Forms;
+using OpenTabletDriver.Plugin;
+
+namespace OpenTabletDriver.UX.RPC
+{
+    public class UXDriver : IUXDriver
```
Comment (InfinityGhost) [inline OpenTabletDriver.UX/RPC/UXDriver.cs:7]: `UserInterfaceProxy`

File: OpenTabletDriver.UX/UXInstance.cs [hunk 3]
```diff
@@ -0,0 +1,54 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using OpenTabletDriver.Desktop.RPC;
+using OpenTabletDriver.Plugin;
+using OpenTabletDriver.UX.RPC;
+
+namespace OpenTabletDriver.UX
+{
+    public class UXInstance : IDisposable
```
Comment (InfinityGhost) [inline OpenTabletDriver.UX/UXInstance.cs:10]: `UserInterfaceInstance`

File: OpenTabletDriver.UX/RPC/UXDriver.cs [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+        public async Task ShowClient()
+        {
+            await Application.Instance.InvokeAsync(() =>
+            {
+                if (Application.Instance?.MainForm is Form form)
+                {
+                    form.Show();
+                    form.WindowState = WindowState.Normal;
+                    form.BringToFront();
+                    form.WindowStyle = WindowStyle.Default;
```
Comment (InfinityGhost) [inline OpenTabletDriver.UX/RPC/UXDriver.cs:18]: This shouldn't be necessary, and actually has the potential to cause issues on non MainForm windows.

File: OpenTabletDriver.UX/UXInstance.cs [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+            if (!Instance.AlreadyExists)
+            {
+                Host = new RpcHost<UXDriver>("OpenTabletDriver.UX");
+                IsClient = false;
+            }
+            else
+            {
+                Client = new RpcClient<IUXDriver>("OpenTabletDriver.UX");
+                IsClient = true;
+            }
```
Comment (InfinityGhost) [inline OpenTabletDriver.UX/UXInstance.cs:28]: Why don't we just null check to see if it is a client or host?

File: OpenTabletDriver.UX/UXInstance.cs [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+using OpenTabletDriver.UX.RPC;
+
+namespace OpenTabletDriver.UX
+{
+    public class UXInstance : IDisposable
+    {
+        readonly Instance Instance = new Instance("OpenTabletDriver.UX.Instance");
+        readonly RpcHost<UXDriver> Host;
+        readonly RpcClient<IUXDriver> Client;
+        readonly bool IsClient;
```
Comment (InfinityGhost) [inline OpenTabletDriver.UX/UXInstance.cs:15]: Add `private` to all fields, use `camelCase` for naming.

**Review by InfinityGhost — CHANGES_REQUESTED** [review]

File: OpenTabletDriver.UX/RPC/UXDriver.cs (hunk 4 above)
Comment (X9VoiD) [reply OpenTabletDriver.UX/RPC/UXDriver.cs:18]: So just keep this?
```cs
                    form.Show();
                    form.BringToFront();
```
I'm assuming we won't need WindowState too to respect user-set window state (fullscreen vs windowed)

Comment (InfinityGhost) [reply OpenTabletDriver.UX/RPC/UXDriver.cs:18]: Yes

**Comment by X9VoiD** [comment]
Requested changes pushed

File: OpenTabletDriver.UX/RPC/UserInterfaceProxy.cs [hunk 7]
```diff
@@ -0,0 +1,21 @@
+using System.Threading.Tasks;
+using Eto.Forms;
+using OpenTabletDriver.Plugin;
+
+namespace OpenTabletDriver.UX.RPC
+{
+    public class UserInterfaceProxy : IUserInterface
```
Comment (InfinityGhost) [inline OpenTabletDriver.UX/RPC/UserInterfaceProxy.cs:7]: Actually, we should make the MainForm implement this interface and pass that through RPC

File: OpenTabletDriver.UX/UserInterfaceInstance.cs [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+                    IsBackground = true
+                };
+                thread.Start();
+            }
+        }
+
+        public void Dispose()
+        {
+            instance?.Dispose();
+            GC.SuppressFinalize(this);
```
Comment (InfinityGhost) [inline OpenTabletDriver.UX/UserInterfaceInstance.cs:44]: All references should be nulled before suppressing finalize (host and client)

File: OpenTabletDriver.UX/UserInterfaceInstance.cs [hunk 9]
```diff
@@ -0,0 +1,47 @@
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+using OpenTabletDriver.Desktop.RPC;
+using OpenTabletDriver.Plugin;
+using OpenTabletDriver.UX.RPC;
+
+namespace OpenTabletDriver.UX
+{
+    public class UserInterfaceInstance : IDisposable
```
Comment (InfinityGhost) [inline OpenTabletDriver.UX/UserInterfaceInstance.cs:10]: This should be `UserInterfaceProxy` instead if we do the MainForm implementation

**Review by InfinityGhost — CHANGES_REQUESTED** [review]

**Comment by X9VoiD** [comment]
Pushed, made some changes to `RpcHost` to allow passing an object the rpc will attach to.

File: OpenTabletDriver.UX/App.cs [hunk 10]
```diff
@@ -10,13 +10,21 @@
...
 using OpenTabletDriver.Plugin;
+using OpenTabletDriver.UX.RPC;
 
 namespace OpenTabletDriver.UX
 {
     public static class App
     {
         public static void Run(string platform, string[] args)
         {
+            _ = UserInterfaceProxy.Invoke(async ux =>
```
Comment (InfinityGhost) [inline OpenTabletDriver.UX/App.cs:21]: Any reason to not just make this method `async void`?

Comment (X9VoiD) [reply OpenTabletDriver.UX/App.cs:21]: No real reason other than the internet saying `async void` might be bad

Comment (X9VoiD) [reply OpenTabletDriver.UX/App.cs]: Changed to `async void`

**Comment by EXtremeExploit** [comment]
no merge?
⠀⣞⢽⢪⢣⢣⢣⢫⡺⡵⣝⡮⣗⢷⢽⢽⢽⣮⡷⡽⣜⣜⢮⢺⣜⢷⢽⢝⡽⣝
⠸⡸⠜⠕⠕⠁⢁⢇⢏⢽⢺⣪⡳⡝⣎⣏⢯⢞⡿⣟⣷⣳⢯⡷⣽⢽⢯⣳⣫⠇
⠀⠀⢀⢀⢄⢬⢪⡪⡎⣆⡈⠚⠜⠕⠇⠗⠝⢕⢯⢫⣞⣯⣿⣻⡽⣏⢗⣗⠏⠀
⠀⠪⡪⡪⣪⢪⢺⢸⢢⢓⢆⢤⢀⠀⠀⠀⠀⠈⢊⢞⡾⣿⡯⣏⢮⠷⠁⠀⠀
⠀⠀⠀⠈⠊⠆⡃⠕⢕⢇⢇⢇⢇⢇⢏⢎⢎⢆⢄⠀⢑⣽⣿⢝⠲⠉⠀⠀⠀⠀
⠀⠀⠀⠀⠀⡿⠂⠠⠀⡇⢇⠕⢈⣀⠀⠁⠡⠣⡣⡫⣂⣿⠯⢪⠰⠂⠀⠀⠀⠀
⠀⠀⠀⠀⡦⡙⡂⢀⢤⢣⠣⡈⣾⡃⠠⠄⠀⡄⢱⣌⣶⢏⢊⠂⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⢝⡲⣜⡮⡏⢎⢌⢂⠙⠢⠐⢀⢘⢵⣽⣿⡿⠁⠁⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠨⣺⡺⡕⡕⡱⡑⡆⡕⡅⡕⡜⡼⢽⡻⠏⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⣼⣳⣫⣾⣵⣗⡵⡱⡡⢣⢑⢕⢜⢕⡝⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⣴⣿⣾⣿⣿⣿⡿⡽⡑⢌⠪⡢⡣⣣⡟⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⡟⡾⣿⢿⢿⢵⣽⣾⣼⣘⢸⢸⣞⡟⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀⠀
⠀⠀⠀⠀⠁⠇⠡⠩⡫⢿⣝⡻⡮⣒⢽⠋⠀⠀⠀⠀⠀⠀

**Comment by doubledge-spec** [comment]
Merge?

**Comment by X9VoiD** [comment]
Updated the thing

**Comment by doubledge-spec** [comment]
update? what are the plans for this after DI rewrite is complete?

**Comment by X9VoiD** [comment]
Superseded by #2828

**Pull request closed without merging**